staging/src/k8s.io/apiserver/pkg/server/options/encryption_provider_config_test.go

```diff
@@ -0,0 +1,102 @@
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package options
+
+import (
+	"strings"
+	"testing"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/storage/value"
+)
+
+var correctConfig string = `
+- kind: k8s-aes-gcm
+  version: v1
+  keys:
+    - name: key1
+      secret: c2VjcmV0IGlzIHNlY3VyZQ==
+    - name: key2
+      secret: dGhpcyBpcyBwYXNzd29yZA==
+  resource: /registry/namespaces
+- kind: k8s-aes-gcm
+  version: v1
+  keys:
+    - name: key2
+      secret: dGhpcyBpcyBwYXNzd29yZA==
+    - name: key3
+      secret: azhzIHNlY3JldCBzdG9yZQ==
+  resource: namespaces,secrets
+`
+
+var incorrectConfigNoSecretForKey string = `
+- kind: k8s-aes-gcm
+  version: v1
+  keys:
+    - name: key1
+    - name: key2
+      secret: dGhpcyBpcyBwYXNzd29yZA==
+  resource: namespaces,secrets
+`
+
+var incorrectConfigNoResource string = `
+- kind: k8s-aes-gcm
+  version: v1
+  keys:
+    - name: key2
+      secret: dGhpcyBpcyBwYXNzd29yZA==
+`
+
+var incorrectConfigInvalidKey string = `
+- kind: k8s-aes-gcm
+  version: v1
+  keys:
+    - name: key1
+      secret: YSBzZWNyZXQgYSBzZWNyZXQ=
+  resource: namespaces,secrets
+`
+
+func TestEncryptionProviderConfigCorrect(t *testing.T) {
+	var destination map[schema.GroupResource]value.Transformer
+	if err := ConfigToTransformerOverrides(strings.NewReader(correctConfig), &destination); err != nil {
+		t.Fatalf("error while parsing configuration file: %s", err)
+	}
+}
+
+// Throw error if key has no secret
+func TestEncryptionProviderConfigNoSecretForKey(t *testing.T) {
+	var destination map[schema.GroupResource]value.Transformer
+	if ConfigToTransformerOverrides(strings.NewReader(incorrectConfigNoSecretForKey), &destination) == nil {
+		t.Fatalf("invalid configuration file (one key has no secret) got parsed:\n%s", incorrectConfigNoSecretForKey)
+	}
+}
+
+// Throw error if provider has no resource
+func TestEncryptionProviderConfigNoResource(t *testing.T) {
+	var destination map[schema.GroupResource]value.Transformer
+	if ConfigToTransformerOverrides(strings.NewReader(incorrectConfigNoResource), &destination) == nil {
+		t.Fatalf("invalid configuration file (one provider has no resource) got parsed:\n%s", incorrectConfigNoResource)
+	}
+}
+
+// Throw error if invalid key for AES
+func TestEncryptionProviderConfigInvalidKey(t *testing.T) {
+	var destination map[schema.GroupResource]value.Transformer
+	if ConfigToTransformerOverrides(strings.NewReader(incorrectConfigInvalidKey), &destination) == nil {
```

Comment:
The reason the parsing had been done manually was that we need to accommodate different types of configurations. For instance, we know that the AES transformer needs a list of keys. The KMS transformer shall need an api-token. Hashicorp vault may need some different information. We needed to parse the YAML on the basis of the `kind:` key's value.

Instead, we came up with an alternative configuration:
```
- k8s-aes-v1:
    keys:
      - name: key1
        secret: c2VjcmV0IGlzIHNlY3VyZQ==
  resource: namespaces
- k8s-kms-v1:
    apitoken: foobar
  resource: secrets
```

We can parse a struct:
```
type Config struct {
    aesConfig AesConfig `yaml:"k8s-aes-v1"`
    kmsConfig KmsConfig `yaml:"k8s-kms-v1"`
    vaultConfig VaultConfig `yaml:"k8s-vault-v1"`
    resource string `yaml:"resource"`
}
```
And later check which one of them was not nil. The idea is to be able to parse the configuration differently on the basis of a particular property.

Another benefit of this one is that we can now keep AesConfig, KmsConfig and other such structs along with the code for the transformer, which hides the configuration details of the transformers from the upper layers. The versioning can be easily handled by the yaml key name as well.

I tried a prototype of the above parsing. Seems feasible.